### PR TITLE
Add bundler Dependency to gemspec

### PR DESCRIPTION
**Pull request by sonalkr132**
`bundle exec` was added to acceptance test cause otherwise it was giving Gem::LoadError for conflicting bundler versions.

An alternate solution would have been to add it [here](https://github.com/rails/spring/blob/master/lib/spring/test/application.rb#L94-L103):

```Ruby
 Bundler.with_clean_env do
   Process.spawn(
     env,
     command.to_s,   # may be change this line to "bundle exec #{command}"
     out:   stdout.last,
     err:   stderr.last,
     in:    :close,
     chdir: root.to_s,
   )
 end
```
I felt it was non-intuitive. Additionally, above code block doesn't get called only from `Spring::Test::AcceptanceTest`, so it would need changes on many different files.

**EDIT:**
Resolves: #275

### CODE REVIEW

File: lib/spring/test/acceptance_test.rb [hunk 1]
```diff
@@ -83,14 +83,14 @@ def without_gem(name)
       end
 
       test "help message when called without arguments" do
-        assert_success "bin/spring", stdout: 'Usage: spring COMMAND [ARGS]'
+        assert_success "bundle exec bin/spring", stdout: 'Usage: spring COMMAND [ARGS]'
```
Comment (grosser) [inline lib/spring/test/acceptance_test.rb]: we always advertise to not use bundle exec ... so with_clean_env would be better :(

Comment (sonalkr132) [inline lib/spring/test/acceptance_test.rb]: I think error on travis (check [this](https://travis-ci.org/sonalkr132/spring/jobs/95848491)) can be on user side as well (?). Looking at the test won't help him a lot cause it won't be apparent that command passing in test is not actually:
```Ruby
bin/spring
```
but 
```Ruby
bundle exec bin/spring
```

Also, **some** of the commands being passed to `with_clean_env` are:
```Ruby
# gives syntax error near unexpected token `('
# can be solved by just removing '('
(gem list bundler | grep bundler) || gem install bundler
# not sure from where does this gets passed some jugglery will be needed    
bundle check || bundle update --retry=2 

## some gem install commands
## more commands from other files calling the same app.run 
```

Comment (sonalkr132) [inline lib/spring/test/acceptance_test.rb]: @grosser Do you really want me to change it to `with_clean_env`?
Actually, it needs just 2 lines of change.

Comment (grosser) [inline lib/spring/test/acceptance_test.rb]: I'm confused now :D 
Do you think this is better or with_clean_env is better ?
I think either would work, I was just sad to see bundle exec being added all over the place + that it uses a pattern we try to discourage ...

Comment (sonalkr132) [inline lib/spring/test/acceptance_test.rb]: `with_clean_env` feels like we are secretly adding `bundle exec` to every command.
However, `with_clean_env` is lot cleaner and gets the job done. Also, issue of conflicting bundler version seems to be only specific for travis runs. Locally, I have bundler 1.10.6 and I don't get any error.

**Comment by sonalkr132** [comment]
@grosser I changed it to `with_clean_env` version. Can you please have a look?
Order of `application.bundle` and `build_and_install_gems` is changed cause to bundle and install gems we will need bundler first. Bundler is install by `application.bundle`.

**Comment by grosser** [comment]
looks good to me ... but could break all kinds of stuff ... @jonleighton want to have a look ?

**Comment by sonalkr132** [comment]
> but could break all kinds of stuff 

Are referring to change of order of `application.bundle` and `build_and_install_gems`? Those things are only used for tests and has nothing to do with working of gem.

**Comment by grosser** [comment]
ahhh you are right! :D 
so you don't need a new release after this ,right ?

**Comment by sonalkr132** [comment]
> so you don't need a new release after this ,right?

TBH, I am not sure. It does add `bundler` as hard dependency.

**Comment by sonalkr132** [comment]
Shit! I just noticed it. I did reset hard and forgot to add the most important line -_-

**Comment by grosser** [comment]
:D

**Comment by sonalkr132** [comment]
Fixed it!
> so you don't need a new release after this ,right?

I think it will need changelog that now spring has bundler has hard dependency.

**Comment by grosser** [comment]
:+1: 

@jonleighton / @rafaelfranca looks good ?

File: lib/spring/test/application.rb [hunk 2]
```diff
@@ -200,7 +200,7 @@ def run!(command, options = {})
       end
 
       def bundle
-        run! "(gem list bundler | grep bundler) || gem install bundler", timeout: nil, retry: 2
+        run! "gem list bundler | grep bundler || gem install bundler", timeout: nil, retry: 2
```
Comment (jonleighton) [inline lib/spring/test/application.rb]: Why change this? It makes it less readable IMO

File: spring.gemspec [hunk 3]
```diff
@@ -13,6 +13,7 @@ Gem::Specification.new do |gem|
   gem.files         = Dir["LICENSE.txt", "README.md", "lib/**/*", "bin/*"]
   gem.executables   = gem.files.grep(%r{^bin/}).map{ |f| File.basename(f) }
 
+  gem.add_dependency 'bundler', '~> 1.3'
```
Comment (jonleighton) [inline spring.gemspec]: I think we can make this more permissive. Perhaps `~> 1.0` is fine.

Comment (jonleighton) [inline spring.gemspec]: In fact, `>= 1.0`

File: lib/spring/test/application.rb [hunk 4]
```diff
@@ -94,7 +94,7 @@ def run(command, opts = {})
         Bundler.with_clean_env do
           Process.spawn(
             env,
-            command.to_s,
+            "bundle exec #{command}",
```
Comment (jonleighton) [inline lib/spring/test/application.rb]: Why is this necessary? Which tests fail without it?

We are already setting `GEM_PATH` and `GEM_HOME` in the env of the spawned process, so I don't understand why it would be an issue.

File: lib/spring/test/application.rb (hunk 2 above)
Comment (sonalkr132) [inline lib/spring/test/application.rb]: `bundle exec (gem list bundler | grep bundler) || gem install bundler` gives following error:

> bash: syntax error near unexpected token `('

File: lib/spring/test/application.rb (hunk 4 above)
Comment (sonalkr132) [inline lib/spring/test/application.rb]: > We are already setting GEM_PATH and GEM_HOME in the env of the spawned process, so I don't understand why it would be an issue.

Yes, it shouldn't be an issues and it isn't, at least locally. However, tests on travis fail with conflicting bundler version. For *some* reason [travis installs bundler 1.7.6](https://travis-ci.org/sonalkr132/spring/jobs/96552775#L101) at first and later it ends up conflicting with generated rails app bundler. Same thing happened in last two attempts of this fix ( #293, #373).

**Comment by sonalkr132** [comment]
Can someone please suggest tests so that #456 doesn't happen again? I can't think of anything reasonable.

**Comment by sonalkr132** [comment]
closed in favor of #458
